lib/using_restricted_array.rb

```diff
@@ -1,61 +1,122 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+#                   or the length of the input array
+# ** O stands for worst-case scenario
+# Space complexity: Constant or O(1), because additional space doesn't
+# depend on array size or input
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
+  #raise NotImplementedError
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+# or the length of the input array
+# Space complexity: Constant or O(1), because additional space doesn't
+# depend on array size or input
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+
+  i = 0
+  while i < len # ends before reaching the end
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+# Space complexity: Constant or O(1), because additional space doesn't
+# depend on array size or input
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
+  #raise NotImplementedError
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+# Space complexity: Constant or O(1), because additional space doesn't
+# depend on array size or input
 def find_largest(array, length)
-  raise NotImplementedError
+  max_value = 0
+  length.times do |i|
+    max_value = array[i] if array[i] > max_value
+  end
+  return max_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+# Space complexity: Constant or O(1), because additional space doesn't
+# depend on array size or input
 def find_smallest(array, length)
-  raise NotImplementedError
+  min_value = array[1]
```

Comment:
This variable name is a bit misleading at first glance, but I am able to understand your intent when I read the rest of the code. The only suggestion I am coming up with is setting `min_value = array[0]` and then testing from i = 0 onwards.